db/history.go

```diff
@@ -0,0 +1,211 @@
+	// Timestamp of the last synced envelope.
+	Current time.Time
+	End     time.Time
+
+	RequestID common.Hash
+}
+
+// Key returns unique identifier for this TopicHistory.
+func (t TopicHistory) Key() TopicHistoryKey {
+	key := TopicHistoryKey{}
+	copy(key[:], t.Topic[:])
+	binary.BigEndian.PutUint64(key[4:], uint64(t.Duration))
+	return key
+}
+
+// Value marshalls TopicHistory into bytes.
+func (t TopicHistory) Value() ([]byte, error) {
+	return json.Marshal(t)
+}
+
+// Load TopicHistory from db using key and unmarshalls it.
+func (t *TopicHistory) Load() error {
+	key := t.Key()
+	if (key == TopicHistoryKey{}) {
+		return errors.New("key is empty")
+	}
+	value, err := t.db.Get(key[:])
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(value, t)
+}
+
+// Save persists TopicHistory on disk.
+func (t TopicHistory) Save() error {
+	key := t.Key()
+	val, err := t.Value()
+	if err != nil {
+		return err
+	}
+	return t.db.Put(key[:], val)
+}
+
+// SameRange returns true if topic has same range, which means:
+// true if Current is zero and Duration is the same
+// and true if Current is the same
+func (t TopicHistory) SameRange(other TopicHistory) bool {
+	zero := time.Time{}
+	if t.Current == zero && other.Current == zero {
+		return t.Duration == other.Duration
+	}
+	return t.Current == other.Current
+}
+
+// HistoryRequest is kept in the database while request is in the progress.
+// Stores necessary information to identify topics with associated ranges included in the request.
+type HistoryRequest struct {
+	db      Interface
+	topicDB Interface
+
+	histories []TopicHistory
+
+	// Generated ID
+	ID common.Hash
+	// List of the topics
+	TopicHistoryKeys []TopicHistoryKey
+}
+
+// AddHistory adds instance to internal list of instance and add instance key to the list
+// which will be persisted on disk.
+func (req *HistoryRequest) AddHistory(history TopicHistory) {
+	req.histories = append(req.histories, history)
+	req.TopicHistoryKeys = append(req.TopicHistoryKeys, history.Key())
+}
+
+// Histories returns internal lsit of topic histories.
+func (req *HistoryRequest) Histories() []TopicHistory {
+	// TODO Lazy load from database on first access
+	return req.histories
+}
+
+// Value returns content of HistoryRequest as bytes.
+func (req HistoryRequest) Value() ([]byte, error) {
+	return json.Marshal(req)
+}
+
+// Save persists all attached histories and request itself on the disk.
+func (req HistoryRequest) Save() error {
+	for i := range req.histories {
+		th := &req.histories[i]
+		th.RequestID = req.ID
+		if err := th.Save(); err != nil {
+			return err
+		}
+	}
+	val, err := req.Value()
+	if err != nil {
+		return err
+	}
+	return req.db.Put(req.ID.Bytes(), val)
+}
+
+// Delete HistoryRequest from store and update every topic.
+func (req HistoryRequest) Delete() error {
+	for i := range req.histories {
+		th := &req.histories[i]
+		th.RequestID = common.Hash{}
+		th.Current = th.End
+		if err := th.Save(); err != nil {
+			return err
```

Comment:
so, we don't delete it if any error occures?

also, I would prefer for this struct not to self-destruct and self-load (overwriting it's contents), but rather have these done on some other level, where `Load()` returns an instance of `*HistoryRequest`

Reply:
yeah, so it is safe to not delete a request cause next time we will simply retry request with same topics and updated range

> also, I would prefer for this struct not to self-destruct and self-load (overwriting it's contents), but rather have these done on some other level, where Load() returns an instance of *HistoryRequest

you mean you prefer to update HistoryRequest outside of the Delete method? or something else?

Reply:
it's like I would prefer `Load` and `Delete` to be performed by some struct outside this one

Reply:
but I want too read this PR one more time before making any assumptions

Reply:
i moved topic updates when request is deleted to reactor. but attached histories are still loaded with a request. if it turns out to be a problem i can change that, but for now it makes easier to work with requests